### PR TITLE
Mcast addr ops

### PR DESCRIPTION
**Pull request by ajayparida**

### CODE REVIEW

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 1]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+{
+
+	struct wifi_nrf_vif_ctx_zep *vif_ctx_zep = NULL;
+	struct wifi_nrf_ctx_zep *rpu_ctx_zep = NULL;
+	const struct device *dev = NULL;
+	struct net_eth_addr mac_addr;
+	struct nrf_wifi_umac_mcast_cfg *mcast_info = NULL;
+	enum wifi_nrf_status status;
+
+	dev = net_if_get_device(iface);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:104]: Add a NULL check

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 2]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+	vif_ctx_zep = dev->data;
+	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
+
+	mcast_info = k_calloc(sizeof(*mcast_info), sizeof(char));
+
+	if (!mcast_info) {
+		LOG_ERR("%s: Unable to allocate memory for mcast_info\n", __func__);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:111]: print size

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 3]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+
+		status = wifi_nrf_fmac_set_multicast(rpu_ctx_zep->rpu_ctx,
+						     vif_ctx_zep->vif_idx,
+						     mcast_info);
+
+		if (status == WIFI_NRF_STATUS_FAIL) {
+			LOG_ERR("%s: nrf_wifi_fmac_set_multicast failed\n", __func__);
+		}
+
+	} else if ((mgmt_event == NET_EVENT_IPV6_MADDR_ADD) ||
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:144]: There is lot of common code, we can refactor this

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 4]
```diff
@@ -106,6 +196,27 @@ void wifi_nrf_if_init(struct net_if *iface)
...
+
+	net_mgmt_init_event_callback(&ip_maddr6_add_cb,
+				     ip_maddr_event_handler,
+				     NET_EVENT_IPV6_MADDR_ADD);
+	net_mgmt_add_event_callback(&ip_maddr6_add_cb);
+
+	net_mgmt_init_event_callback(&ip_maddr6_del_cb,
+				     ip_maddr_event_handler,
+				     NET_EVENT_IPV6_MADDR_DEL);
+	net_mgmt_add_event_callback(&ip_maddr6_del_cb);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:219]: You can add them in a single CB handler, by ORing the events 

```
net_mgmt_init_event_callback(&ip_madd_cbs,
                  ip_maddr_event_handler, 
                  NET_EVENT_IPV4_MADDR_ADD |  NET_EVENT_IPV4_MADDR_DEL | 
                   NET_EVENT_IPV6_MADDR_ADD | NET_EVENT_IPV6_MADDR_DEL);
```

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 5]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+		} else {
+			mcast_info->type = MCAST_ADDR_DEL;
+		}
+
+		status = wifi_nrf_fmac_set_multicast(rpu_ctx_zep->rpu_ctx,
+						     vif_ctx_zep->vif_idx,
+						     mcast_info);
+
+		if (status == WIFI_NRF_STATUS_FAIL) {
+			LOG_ERR("%s: nrf_wifi_fmac_set_multicast failed\n", __func__);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:141]: log the MAC address as well

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 6]
```diff
@@ -1082,4 +1082,23 @@ enum wifi_nrf_status wifi_nrf_fmac_conf_ltf_gi(struct wifi_nrf_fmac_dev_ctx *fma
 					       unsigned char he_ltf,
 					       unsigned char he_gi,
 					       unsigned char enabled);
+
+/**
+ * wifi_nrf_fmac_set_multicast - set the multicast filter address.
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1087]: Change multicast to mcast_addr.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 7]
```diff
@@ -1082,4 +1082,23 @@ enum wifi_nrf_status wifi_nrf_fmac_conf_ltf_gi(struct wifi_nrf_fmac_dev_ctx *fma
...
+ * @fmac_dev_ctx: Pointer to the UMAC IF context for a RPU WLAN device.
+ * @if_idx: Index of the interface whose state needs to be changed.
+ * @mcast_info: multicast information to be set
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1090]: multicast should be Multicast

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 8]
```diff
@@ -1082,4 +1082,23 @@ enum wifi_nrf_status wifi_nrf_fmac_conf_ltf_gi(struct wifi_nrf_fmac_dev_ctx *fma
...
+ *
+ * This function is used to send a command (%NRF_WIFI_UMAC_CMD_MCAST_FILTER) to
+ * instruct the firmware to set the multicast filter address with an index of
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1093]: `...with an index of` should be `....to an interface with index` ?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 9]
```diff
@@ -1082,4 +1082,23 @@ enum wifi_nrf_status wifi_nrf_fmac_conf_ltf_gi(struct wifi_nrf_fmac_dev_ctx *fma
...
+ * @if_idx and parameters specified by @mcast_info.
+ *
+ * Returns: Status
+ *              Pass : %WIFI_NRF_STATUS_SUCCESS
+ *              Error: %WIFI_NRF_STATUS_FAIL
+ */
+enum wifi_nrf_status wifi_nrf_fmac_set_multicast(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1100]: Change multicast to mcast_addr

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 10]
```diff
@@ -3329,4 +3329,46 @@ enum wifi_nrf_status wifi_nrf_fmac_twt_teardown(void *dev_ctx,
 
 	return status;
 }
+
+enum wifi_nrf_status wifi_nrf_fmac_set_multicast(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:3333]: Change multicast to mcast_addr

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 11]
```diff
@@ -3329,4 +3329,46 @@ enum wifi_nrf_status wifi_nrf_fmac_twt_teardown(void *dev_ctx,
...
+				 unsigned char if_idx,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:3334]: Indentation

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c [hunk 12]
```diff
@@ -3329,4 +3329,46 @@ enum wifi_nrf_status wifi_nrf_fmac_twt_teardown(void *dev_ctx,
...
+				 struct nrf_wifi_umac_mcast_cfg *mcast_info)
+{
+	enum wifi_nrf_status status = WIFI_NRF_STATUS_FAIL;
+	struct nrf_wifi_umac_cmd_mcast_filter *set_mcast_cmd = NULL;
+
+	unsigned int len = sizeof(*set_mcast_cmd);
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api.c:3340]: Why is extra variable len required...why not just use sizeof(*set_mcast_cmd) where needed?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 13]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
 #endif /* CONFIG_NRF700X_DATA_TX */
 }
 
+static void ip_maddr_event_handler(struct net_mgmt_event_callback *cb,
+				   uint32_t mgmt_event, struct net_if *iface)
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:94]: Put each parameter on a separate line.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 14]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+{
+
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:96]: Unintentional blank line?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c [hunk 15]
```diff
@@ -85,6 +90,91 @@ int wifi_nrf_if_send(const struct device *dev, struct net_pkt *pkt)
...
+	mcast_info = k_calloc(sizeof(*mcast_info), sizeof(char));
+
+	if (!mcast_info) {
+		LOG_ERR("%s: Unable to allocate memory for mcast_info\n", __func__);
+		return;
+	}
+
+	if ((mgmt_event == NET_EVENT_IPV4_MADDR_ADD) ||
+	    (mgmt_event == NET_EVENT_IPV4_MADDR_DEL)) {
+
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:117]: Unintentional blank line?

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c (hunk 4 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c:219]: As discussed offline, the above won't work as we are trying to mix two layers of net_mgmt IPv4 andIPv6, the check in `mgmt_run_callbacks` fails, so, divide in to two handlers ipv4 and ipv6 and you can use multiple events.

**Review by sr1dh48r — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h [hunk 16]
```diff
@@ -1217,12 +1217,28 @@ enum wifi_nrf_status wifi_nrf_fmac_conf_ltf_gi(struct wifi_nrf_fmac_dev_ctx *fma
  * @otp_info: Pointer to the address where the OTP information needs to be copied.
  *
  * This function is used to fetch OTP information from the RPU.
- *
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/inc/fmac_api.h:1220]: unncessary change

**Review by krish2718 — CHANGES_REQUESTED** [review]
Please update RPU patch commit log with details, I did the same in my branch as below but couldn't find any reference to mcast in the log. This is the diff b/w 55 (MAIN) and 59.

```
    drivers: wifi: revB: Update RPU patch to #59
    
    Changelog:
    
     UMAC:
      - Add 802.11d regulatory beacon hints
      - API to configure power save inactivity timer
      - API to retrieve DTIM and beacon interval
      - Set random Nordic OUI MAC address for default VIF (host doesn't send
        this anymore in the init)
    
    Update the interface file as well, these new commands are not supported
    in RevA as that is deprecated.
    
    Fix the header file to use proper macro NRF_WIFI_ETH_ADDR_LEN.
```

**Comment by krish2718** [comment]
My PR https://github.com/nrfconnect/sdk-nrf/pull/9897 also bumps the RPU patch to 59 but with proper commit log, so, you can rebase once that is merged.

**Review by krish2718 — APPROVED** [review]

**Review by sachinthegreen — APPROVED** [review]

**Pull request merged**